docs/adr/0033-optimise-reads-and-writes-from-dynamo-outbox.md.md

```diff
@@ -0,0 +1,91 @@
+* When a collection of messages are deleted from the outbox they're worked through sequentially with separate requests
+
+### Fetching outstanding message count
+
+Every time one or more messages are cleared from the outbox, the `OutboxProducerMediator` checks whether it needs to update it's internal metric for how many outstanding messages currently sit in the outbox, based on when that last check was last run. If it determines a refresh is neccesary, it fetches all outstanding messages from the outbox into memory (but does so asynchronously). This comes with all the queries described above, and if there are a large number of messages outstanding, can eat up a problematic amount of memory. The count of outstanding messages only appears to be used for monitoring purposes.
+
+## Decision
+
+All of the inefficiencies above can be improved with non-breaking changes.
+
+### `OutstandingMessages` operation
+
+Given that the `Outstanding` index is a sparse index, and we wish to pull out the entirety of that index when we perform the operation, this can be a `Scan` operation on the index instead of a `Query`. This removes the need to iterate over topics and shards, and can instead be done as a single HTTP call if the number of outstanding messages allows it (with paging if it doesn't).
+
+The one downside of this is that we cannot specify the ordering of results from a `Scan` operation. If the results are paged, we will not be able to specify that the oldest messages should be retrieved first. Given the performance issues using `Query` operations, and the limitations of Dynamo DB as a storage platform, this feels like a reasonable comprompise to make.
+
+If a specific topic is provided in the `args` dictionary when performing the `OutstandingMessages` operation, then a `Query` would still need to be used, iterating over the shards for that topic.
+
+### `DispatchedMessages` operation
+
+As above - if no topic is provided in the `args` dictionary, use a `Scan` operation to fetch dispatched messages instead of a `Query` operation.
+
+### The `Delivered` GSI
+
+Introduce sharding to the `Delivered` index, using the same number of shards as configured for the `Outstanding` index.
+
+### Getting messages for dispatch
+
+Add overloads of both `Get` and `GetAsync` to the outbox interfaces that take a collection of message IDs instead of just one. For the Dynamo DB implementation, use a `BatchGet` operation to fetch all of them at once. For the other implementations, they can just iterate over their other `Get` methods for now.
+
+If the `BatchGet` operation only returns a subset of the requested messages, throw an exception.
+
+Update `OutboxProducerMediator` to use the new `Get` methods.
+
+### Marking messages as dispatched
+
+When marking a collection of messages as dispatched, us a `BatchWrite` operation to update all of them at once. If any of the updates fail, throw an exception.
```

Comment:
If we fail to mark dispatched in a batch, do we want to throw an exception, or log? In principle, an undispatched message would just be picked up again and resent. The main issue would be increased latency. Perhaps a boolean to indicate if you throw on an partial write, or just log, and the from a Sweeper we would just let it try again next time, but from an immediate write, you would choose by setting an arg